### DISCUSSION TITLE
Server does not want to stop on us_listen_socket_close

### DISCUSSION THREAD

**Discussion by quirm** [uNetworking/uWebSockets#1110] [Q&A] [answered]
Hi, I'm trying to gracefully stop a websocket server but it seems I need to call `close` on every connected client.

A simple example in `main.cpp`:
```cpp
#include "App.h"
#include <iostream>
#include <csignal> // std::signal

struct us_listen_socket_t *gSocket;

int main()
{
  int port = 9001;
  int nonce = 0;

  std::signal(SIGINT, [](int) {
    std::cout << "Closing down the server, press again to kill." << std::endl;
    std::signal(SIGINT, SIG_DFL);
    us_listen_socket_close(0, gSocket);
    std::cout << "Is it dead?" << std::endl;
  });

  struct PerSocketData {
    int id;
  };

  auto app = uWS::App().ws<PerSocketData>("/ws", {
    .open = [&nonce](auto *ws) {
      ws->subscribe("broadcast");
      auto *data = static_cast<PerSocketData *>(ws->getUserData());
      data->id = ++nonce;
      std::cout << "[open] [" << nonce << "]" << std::endl;
    },
    .message = [](auto *ws, std::string_view message, uWS::OpCode opCode) {
      auto *data = static_cast<PerSocketData *>(ws->getUserData());
      std::cout << "[message] [" << data->id << "] " << message << " " << opCode << std::endl;

      if (message == "stop") {
        std::cout << "closing by a client's request...\n";
        us_listen_socket_close(0, gSocket);
        ws->close();
      } else {
        ws->publish("broadcast", message, opCode);
      }
    }
  }).listen(port, [&port](auto *socket) {
    std::cout << "Listening on port " << port << std::endl;
    gSocket = socket;
	}).run();

  std::cout << "Graceful exit..." << std::endl;

  return 0;
}
```

A client that can run from a console in `about:blank` page:
```
const connect = () => {
  const ws = window.ws = new WebSocket('ws://localhost:9001/ws');
  console.log('connecting...');

  ws.addEventListener('open', () => console.log('[connected]'));
  ws.addEventListener('error', () => console.log('[error]'));
  ws.addEventListener('message', evt => console.log('[message]', evt.data));
  ws.addEventListener('close', evt => console.log('[closed]', evt.code, evt.wasClean));
};

connect();
```

Steps:
1. start a server
2. connect with a client
3. interrupt in terminal or call from one client `ws.send('stop')` when two or more clients are connected.

Expected:
The server disconnects all clients and an event loop is stopped.

What is happening:
It just goes through `us_listen_socket_close(0, gSocket);` but the server is still running and all clients are connected. When I disconnect all connected clients, it will stop. Also, after calling `us_listen_socket_close`, I can't connect with a new client which makes sense but I can still keep messaging with currently active clients and the server is publishing messages to all participants.

Could you please point me to the right API on how to gracefully stop the server or whether I have to manually keep all connected clients and call close on every one of them? Thank you

**Comment by ghost** [comment, marked as answer]
us_listen_socket_close shuts the listen socket down. All other established sockets are unaffected. So you need to close them (or gracefully end them) yourself.

It had been brought up that we could add us_socket_context_close as a way to close all sockets in a context, but I haven't added it since it is quite easy to do this on your own.

**Reply by ElectroluxV2** [reply]
Could you post implementation of us_socket_context_close?

**Comment by quirm** [comment]
It would be _nice-to-have_ in cases where everything can be handled just through subscription. Thank you for the clarification.

**Reply by z16166** [reply]
1.How to close all client sockets from CTRL + C handler routine?
2. Can we safely call us_listen_socket_close() from the  CTRL + C handler thread to close the listening socket while the listening thread is using it? no race there? I randomly get access violation inside us_listen_socket_close().

**Reply by ghost** [reply]
Du kan ta och lista ut saker själv med tanke på din stil
